### ISSUE TITLE
[SPIKE] Investigate Windows binary options

### ISSUE THREAD

**Issue by maxrake** [phylum-dev/phylum-ci#472]
## Description

This is a research spike to investigate whether it is possible to create a standalone Windows binary for the `phylum-ci` package.

## Additional Details

* Functionality from the `phylum-ci` entry point is desired to be in the binary
  * It would be okay if the `phylum-init` entry point is not included
* Creating a standalone binary is the primary goal
* Creating a single file standalone binary is the stretch goal
* Test with the recently added Phylum CLI Windows binary
  * https://github.com/phylum-dev/cli/pull/1486
  * Assume this binary will be present in the test environment and on the PATH
* Possible solutions
  * [Beeware Briefcase](https://briefcase.readthedocs.io/en/stable/)
  * [py2exe](https://www.py2exe.org/)
  * [PyInstaller](https://pyinstaller.org/en/stable/)
  * [Nuitka](https://nuitka.net/user-documentation/)

## Acceptance Criteria

- [x] Determine whether standalone binary creation is possible
- [x] Test the produced binary in a representative environment
- [x] Document any idiosyncrasies

**Comment by maxrake** [comment]
It is possible to create a single file standalone Windows binary that has the same functionality offered in the `phylum-ci` script entry point!

The binary can be viewed [here, in a recent `Preview` workflow run](https://github.com/phylum-dev/phylum-ci/actions/runs/10886437785). It can also be generated with a manual run of the `Preview` workflow, using the `windows_check` branch and selecting the option to "Create Windows binary."

Nuitka was used to generate the binary since it appeared to have the most features and active development. Beeware's Briefcase also looks very polished but it generates MSI installer files, which would require an extra step compared to a standalone EXE file. The `py2exe` tool appears a bit dated and lacks some of the universal compatibility/portability support that Nuitka provides. The `PyInstaller` tool appears to be the next best choice. It may have also met the requirements but was not explored in detail since `Nuitka` worked.

Testing with the produced binary in a representative environment is still needed.

**Comment by maxrake** [comment]
Initial testing is showing some rough edges with the way the Phylum CLI treats Windows canonicalized paths, namely in the `.phylum_project` file. When such a file is present, and the dependency files are acquired from it, they include the Windows device drive prefix `\\?\`. This causes an unhandled exception due to using `os.path.relpath` for debug printable string representations:

<details><summary>Details</summary>
<p>

```
INFO     Using Phylum group: phylum_bot                                        
Exception in thread Thread-14 (_readerthread):
Traceback (most recent call last):
  File "C:\Users\RUNNER~1\AppData\Local\PHYLUM~1\049~1.1-4\threading.py", line 1075, in _bootstrap_inner
  File "C:\Users\RUNNER~1\AppData\Local\PHYLUM~1\049~1.1-4\threading.py", line 1012, in run
  File "C:\Users\RUNNER~1\AppData\Local\PHYLUM~1\049~1.1-4\subprocess.py", line 1599, in _readerthread
  File "encodings\cp1252.py", line 23, in decode
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 1: character maps to <undefined>
INFO     Project/group pairing already exists. Continuing with it.             
             Project: phylum-ci                                                
             Group:   phylum_bot                                               
DEBUG    Repository URL not available to set                                   

INFO     No valid dependency files were provided as arguments. An attempt will 
         be made to detect them.                                               
+--------------------- Traceback (most recent call last) ---------------------+
| in <module>:271                                                             |
|                                                                             |
| in script_main:267                                                          |
|                                                                             |
| in main:230                                                                 |
|                                                                             |
| in __get__:993                                                              |
|                                                                             |
| in depfiles:[165](https://github.com/phylum-dev/phylum-ci/actions/runs/10914352520/job/30292353042#step:9:166)                                                             |
|                                                                             |
| in debug:1527                                                               |
|                                                                             |
| in _log:[168](https://github.com/phylum-dev/phylum-ci/actions/runs/10914352520/job/30292353042#step:9:169)4                                                                |
|                                                                             |
| in handle:[170](https://github.com/phylum-dev/phylum-ci/actions/runs/10914352520/job/30292353042#step:9:171)0                                                              |
|                                                                             |
| in callHandlers:1762                                                        |
|                                                                             |
| in handle:1028                                                              |
|                                                                             |
| in emit:134                                                                 |
|                                                                             |
| in format:999                                                               |
|                                                                             |
| in format:703                                                               |
|                                                                             |
| in getMessage:392                                                           |
|                                                                             |
| in __repr__:66                                                              |
|                                                                             |
| in relpath:783                                                              |
+-----------------------------------------------------------------------------+
ValueError: path is on mount '\\\\?\\D:', start on mount 'D:'
```

</p>
</details>

The full logs can be [viewed here](https://github.com/phylum-dev/phylum-ci/actions/runs/10914352520/job/30292353042). Work is progressing to determine a good path forward. Possible options:

* Bypass the exception and make do with uglier debug output
* Address the canonicalization issue in Phylum CLI, [like was done recently](https://github.com/phylum-dev/cli/pull/1496)

**Comment by maxrake** [comment]
The `\\?\` prefix issue has been resolved with a tested solution in the Phylum CLI, by using the `dunce` crate for paths in the phylum project file. A separate review for that will go up shortly.

**Comment by maxrake** [comment]
There was another issue where the v7.0.0 release of Phylum CLI does not enable extensions on the Windows binary. Extensions are possible there, by enabling them as a feature when building. A separate review will go up for that. Testing with a CLI binary with extensions enabled showed a successful run. The test cases were against the `phylum-ci` repository:

* `phylum-ci -vvaf`
  * ✅ Default checkout
  * ✅ Renaming the `.phylum_project` file so that automatic detection is used
  * ✅ Adding a `build.gradle` file to see that it gets recognized
* `phylum-ci -vvaf --group test --exclude "build.gradle"`
  * ✅ Adding a `build.gradle` file to see that it gets properly ignored; group/project pairing was created and used

**Comment by maxrake** [comment]
One remaining idiosyncrasy is the following output, mixed in with the standard output:

```
Exception in thread Thread-14 (_readerthread):
Traceback (most recent call last):
  File "C:\Users\RUNNER~1\AppData\Local\PHYLUM~1\049~1.1-4\threading.py", line 1075, in _bootstrap_inner
  File "C:\Users\RUNNER~1\AppData\Local\PHYLUM~1\049~1.1-4\threading.py", line 1012, in run
  File "C:\Users\RUNNER~1\AppData\Local\PHYLUM~1\049~1.1-4\subprocess.py", line 1599, in _readerthread
  File "encodings\cp1252.py", line 23, in decode
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 1: character maps to <undefined>
```

This looks like a file being read with the default Windows cp1252 encoding when it should be something else...perhaps UTF-8. The file is likely `.phylum_project`.

FWIW, it doesn't appear that the "errors" hamper execution of the program...but it is not a good look in the output. Further investigation and testing will come next.

**Comment by kylewillmon** [comment]
> The file is likely `.phylum_project`.

Why would there be an 0x9d byte in `.phylum_project`? Are you testing with a non-ascii path or group name or something?

**Comment by maxrake** [comment]
> Why would there be an 0x9d byte in `.phylum_project`? Are you testing with a non-ascii path or group name or something?

There isn't...that was just an initial WAG. Further inspection of the codebase reveals that all file reads are done with an encoding specified. I'm still not sure where this originates. Maybe it comes from a file or bootstrap code that Nuitka generates. I'll look a little further, perhaps with some SysInternals tools that can better pinpoint the resource accessed when this exception occurs.

**Comment by kylewillmon** [comment]
Looking back at your traceback, I think this is coming from a call to `subprocess.run()` and not a file read... although I'm still not sure what program would be outputting 0x9d

**Comment by maxrake** [comment]
> The `\\?\` prefix issue has been resolved with a tested solution in the Phylum CLI, by using the `dunce` crate for paths in the phylum project file. A separate review for that will go up shortly.

https://github.com/phylum-dev/cli/pull/1500 was created for this.

**Comment by maxrake** [comment]
The next, and hopefully last, issue to address is a failure during analysis:

<details><summary>Details (click to expand...)</summary>
<p>

```
DEBUG    71 unique current dependencies from 1 file(s)                         
INFO     Considering all current dependencies ...                              
INFO     Performing analysis. This may take a few seconds.                     
DEBUG    Using analysis command: phylum.EXE extension run --yes                
         'C:\Users\runneradmin\AppData\Local\phylum-ci\0.49.1-454\phylum\exts\c
         i' phylum-ci No-CI_windows_check_9a3d09f --group phylum_bot           
         'C:\Users\RUNNER~1\AppData\Local\Temp\base_g0ld1rs_.json'             
         'C:\Users\RUNNER~1\AppData\Local\Temp\curr_zb_b26ov.json'             

+----------------------------- Subprocess Error ------------------------------+
| COMMAND: phylum.EXE extension run --yes                                     |
| C:\Users\runneradmin\AppData\Local\phylum-ci\0.49.1-454\phylum\exts\ci      |
| phylum-ci No-CI_windows_check_9a3d09f --group phylum_bot                    |
| C:\Users\RUNNER~1\AppData\Local\Temp\base_g0ld1rs_.json                     |
| C:\Users\RUNNER~1\AppData\Local\Temp\curr_zb_b26ov.json                     |
| RETURN CODE: 1                                                              |
| +-------------------------------- STDERR ---------------------------------+ |
| | \u2757 Error: Requires read access to                                   | |
| | "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\curr_zb_b26ov.json"         | |
| +-------------------------------------------------------------------------+ |
+-----------------------------------------------------------------------------+
CRITICAL There was a problem analyzing the project.                            
         A paid account is needed to use groups: https://phylum.io/pricing     
         If the command was expected to succeed, please report this as a bug.  
```

</p>
</details>

This failure appears to be due to an inability to access a named temporary file used during the analysis command. The access is attempted in the `ci` Phylum CLI extension, during a `Deno.readTextFile(current)` call:

https://github.com/phylum-dev/phylum-ci/blob/c0383f085acfce544128f584bc6d1a038ace33b2/src/phylum/exts/ci/main.ts#L27

The odd thing is that this failure does not occur when testing on a local Windows system. It is almost like the `windows-latest` GitHub-hosted runner where this failure occurs is limiting the ability to read files from the temporary directory...but that doesn't quite check out given the file is allowed to be written there.

The temporary files are created with `tempfile.NamedTemporaryFile` here:

https://github.com/phylum-dev/phylum-ci/blob/30b1ce34214c9a4ff41dff40e8c868232f777960/src/phylum/ci/ci_base.py#L801

[Documentation for this call](https://devdocs.io/python~3.12/library/tempfile) indicates the following:

> If delete is true (the default) and delete_on_close is true (the default), the file is deleted as soon as it is closed. If delete is true and delete_on_close is false, the file is deleted on context manager exit only, or else when the [file-like object](https://devdocs.io/python~3.12/glossary#term-file-like-object) is finalized. Deletion is not always guaranteed in this case (see [object.__del__()](https://devdocs.io/python~3.12/reference/datamodel#object.__del__)). If delete is false, the value of delete_on_close is ignored.
> 
> Therefore to use the name of the temporary file to reopen the file after closing it, either make sure not to delete the file upon closure (set the delete parameter to be false) or, in case the temporary file is created in a [with](https://devdocs.io/python~3.12/reference/compound_stmts#with) statement, set the delete_on_close parameter to be false. The latter approach is recommended as it provides assistance in automatic cleaning of the temporary file upon the context manager exit.
>
> Opening the temporary file again by its name while it is still open works as follows:
> 
> * On POSIX the file can always be opened again.
> * On Windows, make sure that at least one of the following conditions are fulfilled:
>   * delete is false
>   * additional open shares delete access (e.g. by calling [os.open()](https://devdocs.io/python~3.12/library/os#os.open) with the flag O_TEMPORARY)
>   * delete is true but delete_on_close is false. Note, that in this case the additional opens that do not share delete access (e.g. created via builtin [open()](https://devdocs.io/python~3.12/library/functions#open)) must be closed before exiting the context manager, else the [os.unlink()](https://devdocs.io/python~3.12/library/os#os.unlink) call on context manager exit will fail with a [PermissionError](https://devdocs.io/python~3.12/library/exceptions#PermissionError).
>
> On Windows, if delete_on_close is false, and the file is created in a directory for which the user lacks delete access, then the [os.unlink()](https://devdocs.io/python~3.12/library/os#os.unlink) call on exit of the context manager will fail with a [PermissionError](https://devdocs.io/python~3.12/library/exceptions#PermissionError). This cannot happen when delete_on_close is true because delete access is requested by the open, which fails immediately if the requested access is not granted.

So...the next attempt will be to try with `delete` as true and `delete_on_close` as false.

**Comment by maxrake** [comment]
The [latest run](https://github.com/phylum-dev/phylum-ci/actions/runs/11016156691/job/30591024519) resulted in this error:

```
+--------------------- Traceback (most recent call last) ---------------------+
| in <module>:271                                                             |
|                                                                             |
| in script_main:267                                                          |
|                                                                             |
| in main:247                                                                 |
|                                                                             |
| in traced_function:236                                                      |
|                                                                             |
| in analyze:801                                                              |
+-----------------------------------------------------------------------------+
TypeError: NamedTemporaryFile() got an unexpected keyword argument 'delete_on_close'
```

I think that means the binary was compiled as something other than Python 3.12, since that is the version when the `delete_on_close` parameter was added.

**Comment by maxrake** [comment]
I was able to get it working by:

* Building the binary with Python 3.12
* Changing the `read` permission in `src/phylum/exts/ci/PhylumExt.toml` from `["/"]` to `true`

Now, to see if the `delete_on_close` is truly needed because it is only available for Python 3.12+

**Comment by maxrake** [comment]
> Now, to see if the `delete_on_close` is truly needed because it is only available for Python 3.12+

It worked without `delete_on_close`! That takes care of the last remaining issue. There are no idiosyncrasies left to disclose. More robust local and CI-based testing is likely needed but the thought is that this spike ticket is ready for acceptance testing and the results can be applied to a new `phylum-ci` release.
